src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,532 @@
+        in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] = f_uncore_efficient;
+
+        // Validate all (uncore frequency, max memory bandwidth) pairs
+        std::set<double> policy_uncore_freqs;
+        for (auto it = in_policy.begin() + M_POLICY_FIRST_UNCORE_FREQ;
+             it != in_policy.end() && std::next(it) != in_policy.end(); std::advance(it, 2)) {
+            auto mapped_mem_bw = *(it + 1);
+            if (!std::isnan(*it)) {
+                // We are using a static cast rather than reinterpreting the
+                // memory so that regions can be input to this policy in the
+                // same form they are output from a report.
+                auto uncore_freq = static_cast<uint64_t>(*it);
+                if (std::isnan(mapped_mem_bw)) {
+                    throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                    "(): mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth.",
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+                // Just make sure the frequency does not have multiple definitions.
+                if (!policy_uncore_freqs.insert(uncore_freq).second) {
+                    throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                    " policy has multiple entries for CPU_UNCORE_FREQUENCY " +
+                                    std::to_string(uncore_freq),
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+            }
+            else if (!std::isnan(mapped_mem_bw)) {
+                throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                " policy maps a NaN CPU_UNCORE_FREQUENCY with max memory bandwidth: " +
+                                std::to_string(mapped_mem_bw),
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+
+            }
+        }
+
+    }
+
+    // Distribute incoming policy to children
+    void CPUActivityAgent::split_policy(const std::vector<double>& in_policy,
+                                    std::vector<std::vector<double> >& out_policy)
+    {
+        for (auto &child_pol : out_policy) {
+            child_pol = in_policy;
+        }
+    }
+
+    // Indicate whether to send the policy down to children
+    bool CPUActivityAgent::do_send_policy(void) const
+    {
+        return true;
+    }
+
+    void CPUActivityAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                        std::vector<double>& out_sample)
+    {
+
+    }
+
+    // Indicate whether to send samples up to the parent
+    bool CPUActivityAgent::do_send_sample(void) const
+    {
+        return false;
+    }
+
+    void CPUActivityAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        m_do_write_batch = false;
+
+        if (m_update_qm_max_rate) {
```

Comment:
The body of this if() is only evaluated on the first call to ```adjust_platform()```.  ```m_update_qm_max_rate``` is set to true in the initializer list, and false at the end of this if().  Was this intentional?

Reply:
This was intentional.  The system characterization used by the agent in this version is based upon a single system wide characterization provided via policy and could be handled in the validate_policy section, but future updates in PR #2437 allow it to use per node signals (added in PR #2435) which necessitated moving the logic into adjust_platform to read the per node signals.